src/Shared/CoreLib/System/Collections/Concurrent/ConcurrentQueueSegment.cs

```diff
@@ -0,0 +1,339 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Collections.Concurrent
+{
+    /// <summary>
+    /// Provides a multi-producer, multi-consumer thread-safe bounded segment.  When the queue is full,
+    /// enqueues fail and return false.  When the queue is empty, dequeues fail and return null.
+    /// These segments are linked together to form the unbounded <see cref="ConcurrentQueue{T}"/>. 
+    /// </summary>
+    [DebuggerDisplay("Capacity = {Capacity}")]
+    internal sealed class ConcurrentQueueSegment<T>
```

Comment:
Why copy this in instead of using the higher-level ConcurrentQueue?

Reply:
ConcurrentQueue is unbounded; whereas the ConcurrentQueueSegment is bounded (and ConcurrentQueue chains them to become unbounded); so it has the characteristics we are after. If we used ConcurrentQueue we'd have to add an additional layer to limit its growth.

Alas exposing it directly was rejected by api review https://github.com/dotnet/corefx/issues/24365

Reply:
That's too bad. How bad do you think the additional layer on ConcurrentQueue would be? I'm thinking TryEnqueue would Interlocked.Increment before Enqueuing into the underlying ConcurrentQueue and TryDequeue would decrement.

While not free, that doesn't sound to bad given one enqueue and dequeue per request. I'm not sure how much ConcurrentQueue itself adds in terms of cost on top of ConcurrentQueueSegment. Hopefully not much.

Reply:
The segment is fine, corefx also has a copy in the codebased used in a bunch of places so it doesn't feel bad copying it into our shared code here.

cc @stephentoub for FYI

Reply:
> corefx also has a copy

Yes and no.  There is a copy of the code, but it's part of the folder that's mirrored between the other repos, where any change in one is automatically propagated to the others.

Reply:
ASP.NET Core will get it a couple weeks later 😄

Reply:
> How expensive do you think the additional layer on ConcurrentQueue would be? I'm thinking TryEnqueue would Interlocked.Increment before Enqueuing into the underlying ConcurrentQueue and TryDequeue would decrement, or vice versa if you like counting down.

I think, it would need to be a compare-and-swap loop to force bounding; increment/decrement would only achieve best endeavours as it wouldn't be atomic with the enqueue/dequeue.

But since @stephentoub is here he can probably provide more insight 😉

Reply:
> > corefx also has a copy

> Yes and no. There is a copy of the code, but it's part of the folder that's mirrored between the other repos, where any change in one is automatically propagated to the others.

Does ASP.NET do any mirroring that could be used? (i.e. for WebSocket; Http/2)

Reply:
When we discussed this for HTTP/2 we decided we'd mirror manually if at all.

Reply:
Also, I really doubt you need to loop to force bounding when you're able to just fail the TryEnqueue call.

Reply:
Could look to see how a ThreadStatic queue per thread would work; would have lower contention than a single count value.

Reply:
This makes me wonder what the threadpool threadcount vs. concurrent request ratio is for a typical web server under load.

I'd hold off and hear what @davidfowl has to think. I'm interpreting "ASP.NET Core will get it a couple weeks later 😄" to mean he has faith in the manual mirroring method.

Reply:
Yes this is fine. Lets keep the queue segment around.